### DISCUSSION TITLE
Sign PDF from server side

### DISCUSSION THREAD

**Discussion by MhmodTayel** [MatthiasValvekens/pyHanko#182] [Q&A]
I'm trying to sign pdf from server side to avoid manipulating timestamp and date of signing by changing the date in PC 

The solution I have in mind is to to separate opening a session with token to sign pdf and creating stamp 
I don't know if it is possible but I didn't find an implementation of what I want till now

**Comment by MatthiasValvekens** [comment]
So, you want to generate the appearance of the signature on the client, but have the server do the signing operations? Or the other way around? Especially when hardware tokens are involved, those are very different (both in implementation and in terms of security implications).

Either way, this

> to avoid manipulating timestamp and date of signing by changing the date in PC

is not really a good reason to change _where_ you do the signing. After all, you'd be replacing one uncorroborated timestamp (from the user's computer) with another equally uncorroborated one (from the server). Looking at the file, there's no way to tell whose clock was used, in the end. The proper way to handle that is by having a time stamping authority witness the signature. PyHanko supports this out of the box; see [the documentation](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#timestamp-handling).

There are many widely trusted time stamping servers that operate at no cost to the user. The one at `http://timestamp.entrust.net/TSS/RFC3161sha2TS` is one of those, for example.

EDIT: converted to discussion.
